[lldb-dap] Add support for data breakpoint.

ZequanWu opened the pull request:
This implements functionality to handle `DataBreakpointInfo` request and `SetDataBreakpoints` request.

If variablesReference is 0 or not provided, interpret name as ${number of bytes}@${expression} to set data breakpoint at the given expression because the spec https://microsoft.github.io/debug-adapter-protocol/specification#Requests_DataBreakpointInfo doesn't say how the client could specify the number of bytes to watch.

This is based on top of https://github.com/llvm/llvm-project/pull/80753.

walter-erquinigo commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/dap_server.py:
```diff
@@ -895,6 +906,32 @@ def request_setFunctionBreakpoints(self, names, condition=None, hitCondition=Non
+        return self.send_recv(command_dict)
+
+    def request_setDataBreakpoint(self, dataBreakpoints):
+        """dataBreakpoints is a list of dictionary with following fields:
```
`list of dictionary` sounds very weird. Did you mean something else?

ZequanWu replied:
`dataBreakpoints` is a list of the dataBreakpoint which has following format:
```
{
    dataId: (address in hex)/(size in bytes)
    accessType: read/write/readWrite
    [condition]: string
    [hitCondition]: string
}
```

walter-erquinigo reviewed (commented):
This looks great overall!
Could you split this PR into two? One with the refactoring and another one with the new features? That would make reviewing easier

ZequanWu commented:
> This looks great overall! Could you split this PR into two? One with the refactoring and another one with the new features? That would make reviewing easier

Yes, I already split it out: https://github.com/llvm/llvm-project/pull/80753 is the refactor pr. I'll rebase this one once that is merged.

walter-erquinigo commented:
just ping me when this PR is rebased

ZequanWu committed:
[lldb-dap] Add support for data breakpoint.

This implements functionality to handle DataBreakpointInfo request and SetDataBreakpoints request.

It doesn't handle the case when name is an expression, see Todo comment for details.

ZequanWu commented:
> just ping me when this PR is rebased

Rebased.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2591,6 +2594,248 @@ void request_setFunctionBreakpoints(const llvm::json::Object &request) {
+  lldb::SBError error;
+  llvm::json::Array accessTypes{"read", "write", "readWrite"};
+  const auto *arguments = request.getObject("arguments");
+  const auto variablesReference =
+      GetUnsigned(arguments, "variablesReference", 0);
```
it'd be better to use a non-zero default value, which might collide with an actual variable ref id

ZequanWu replied:
If `variablesReference` is 0, it's treated as not provided as it's only meaningful if its value > 0. This is also how `request_variables` and `request_setVariable` handle it.

The spec https://microsoft.github.io/debug-adapter-protocol/specification#Types_Variable says: "If `variablesReference` is > 0, the variable is structured ...".

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2591,6 +2594,248 @@ void request_setFunctionBreakpoints(const llvm::json::Object &request) {
+  bool is_duplicated_variable_name = name.contains(" @");
+
+  lldb::SBValue variable;
+  if (lldb::SBValueList *top_scope = GetTopLevelScope(variablesReference)) {
+    // variablesReference is one of our scopes, not an actual variable it is
+    // asking for a variable in locals or globals or registers
+    int64_t end_idx = top_scope->GetSize();
+    // Searching backward so that we choose the variable in closest scope
+    // among variables of the same name.
+    for (int64_t i = end_idx - 1; i >= 0; --i) {
+      lldb::SBValue curr_variable = top_scope->GetValueAtIndex(i);
+      std::string variable_name = CreateUniqueVariableNameForDisplay(
+          curr_variable, is_duplicated_variable_name);
+      if (variable_name == name) {
+        variable = curr_variable;
+        break;
+      }
+    }
+  } else {
+    // We are expanding a variable that has children, so we will return its
+    // children.
+    lldb::SBValue container = g_dap.variables.GetVariable(variablesReference);
+    variable = container.GetChildMemberWithName(name.data());
+    if (!variable.IsValid()) {
+      if (name.starts_with("[")) {
+        llvm::StringRef index_str(name.drop_front(1));
+        uint64_t index = 0;
+        if (!index_str.consumeInteger(0, index)) {
+          if (index_str == "]")
+            variable = container.GetChildAtIndex(index);
+        }
+      }
+    }
+  }
```
This logic is pretty much the same as the one in `request_setVariable`. Could you dedup this code?

ZequanWu replied:
Refactored as a separate function.

walter-erquinigo reviewed (commented):
Amazing stuff! I've been wanting this for a while. I left some minor comments.

ZequanWu committed:
- Address comments.
- Add support for setting data breakpoint with expression. If `variablesReference` is 0 or not provided, interprete `name` as `${number of bytes}@${expression}` to set data breakpoint at the given expression.

ZequanWu commented:
I updated to add support for setting data breakpoint with expression. If `variablesReference` is 0 or not provided, interpret `name` as `${number of bytes}@${expression}` to set data breakpoint at the given expression because the spec https://microsoft.github.io/debug-adapter-protocol/specification#Requests_DataBreakpointInfo doesn't say how the client could specify the number of bytes to watch. But I have concern that using `@` as a separator here, maybe some other languages uses `@` as part of their syntax.

walter-erquinigo commented:
I think using @ is fine, but we can revisit it later if we see any issues.

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -561,6 +561,46 @@ void EventThreadFunction() {
+    // among variables of the same name.
+    for (int64_t i = end_idx - 1; i >= 0; --i) {
+      lldb::SBValue curr_variable = top_scope->GetValueAtIndex(i);
+      std::string variable_name = CreateUniqueVariableNameForDisplay(
+          curr_variable, is_duplicated_variable_name);
```
Do we want to check the name that might include the "foo @ main.cpp:12" or just look for "foo"? Also, some members of a struct be anonymous unions, we probably want to make sure we can do the right thing with those

ZequanWu replied:
This functions is shared between `request_setVariable` and `request_dataBreakpointInfo`, so it checks name that contains `foo @ main.cpp:12`. I tested this manually.

For anonymous unions inside struct, I also tested manually. If we watch either `c` or `d`, changes on either value stops the program as expected.
```
struct A {
  int a;
  int b;
  union {
    char c;
    int d;
  };
};
```

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2697,58 +2737,41 @@ void request_dataBreakpointInfo(const llvm::json::Object &request) {
-      }
-    }
+  if (variable.IsValid()) {
+    addr = llvm::utohexstr(variable.GetLoadAddress());
```
Need some error checking here. `variable.GetLoadAddress()` will only return a valid address if the variable is actually in memory. So if `variable.GetLoadAddress()` returns LLDB_INVALID_ADDRESS we need to return an appropriate error like "<name> does not exist in memory, its location is <location>" where <location> is the result of `const char *SBValue::GetLocation()`.

ZequanWu replied:
Sent a fix at https://github.com/llvm/llvm-project/pull/81680

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2697,58 +2737,41 @@ void request_dataBreakpointInfo(const llvm::json::Object &request) {
-    }
+  if (variable.IsValid()) {
+    addr = llvm::utohexstr(variable.GetLoadAddress());
+    size = llvm::utostr(variable.GetByteSize());
```
Need to check if the byte size is not zero and return an error if it is.

ZequanWu replied:
Sent a fix at https://github.com/llvm/llvm-project/pull/81680

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2697,58 +2737,41 @@ void request_dataBreakpointInfo(const llvm::json::Object &request) {
+    addr = llvm::utohexstr(variable.GetLoadAddress());
+    size = llvm::utostr(variable.GetByteSize());
+  } else if (variablesReference == 0 && frame.IsValid()) {
+    // Name might be an expression. In this case we assume that name is composed
+    // of the number of bytes to watch and expression, separated by '@':
+    // "${size}@${expression}"
```
Is this format VS Code supported, or just something we are making up?

ZequanWu replied:
I'm making this up, since the spec doesn't say how to specify the number of bytes to watch. Do you have better idea?

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -2697,58 +2737,41 @@ void request_dataBreakpointInfo(const llvm::json::Object &request) {
+                                          ? std::string(error_cstr)
+                                          : "evaluation failed");
+    } else
+      addr = llvm::utohexstr(value.GetValueAsUnsigned());
```
We probably want to check this address by making sure it is valid. What if this return zero?

ZequanWu replied:
Sent a fix at https://github.com/llvm/llvm-project/pull/81680

nico commented on llvm/utils/gn/secondary/lldb/tools/lldb-dap/BUILD.gn:
```diff
@@ -52,5 +52,6 @@ executable("lldb-dap") {
     "RunInTerminal.cpp",
     "SourceBreakpoint.cpp",
     "lldb-dap.cpp",
+    "Watchpoint.cpp"
```
Please don't update gn build files if you don't use the gn build. There's a bot that is able to do it most of the time, and it doesn't make mistakes like forgetting the trailing comma here :)

Prabhuk commented:
We (Fuchsia clang toolchain team) are facing a LLDB test failure in our ARM64 Clang toolchain builders and I suspect this change is causing the failure.

Build link: https://ci.chromium.org/raw/build/logs.chromium.org/fuchsia/led/prabhukr_google.com/04517f65babfe0d4ee60f891a290183f9cfe5121932ac926d829bf2f3aee99ea/+/build.proto?server=chromium-swarm.appspot.com

Log: https://logs.chromium.org/logs/fuchsia/led/prabhukr_google.com/04517f65babfe0d4ee60f891a290183f9cfe5121932ac926d829bf2f3aee99ea/+/u/lldb/test/stdout


```
--
Command Output (stderr):
--
FAIL: LLDB (/b/s/w/ir/x/w/llvm_build/bin/clang-aarch64) :: test_expression (TestDAP_setDataBreakpoints.TestDAP_setDataBreakpoints)
FAIL: LLDB (/b/s/w/ir/x/w/llvm_build/bin/clang-aarch64) :: test_functionality (TestDAP_setDataBreakpoints.TestDAP_setDataBreakpoints)
======================================================================
FAIL: test_expression (TestDAP_setDataBreakpoints.TestDAP_setDataBreakpoints)
   Tests setting data breakpoints on expression.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/s/w/ir/x/w/llvm-llvm-project/lldb/test/API/tools/lldb-dap/databreakpoint/TestDAP_setDataBreakpoints.py", line 52, in test_expression
    self.assertEquals(i_val, "2")
AssertionError: '-2320' != '2'
- -2320+ 2
Config=aarch64-/b/s/w/ir/x/w/llvm_build/bin/clang
======================================================================
FAIL: test_functionality (TestDAP_setDataBreakpoints.TestDAP_setDataBreakpoints)
   Tests setting data breakpoints on variable.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/s/w/ir/x/w/llvm-llvm-project/lldb/test/API/tools/lldb-dap/databreakpoint/TestDAP_setDataBreakpoints.py", line 93, in test_functionality
    self.assertEquals(i_val, "2")
AssertionError: '-2320' != '2'
- -2320+ 2
Config=aarch64-/b/s/w/ir/x/w/llvm_build/bin/clang
----------------------------------------------------------------------
Ran 2 tests in 2.339s

RESULT: FAILED (0 passes, 2 failures, 0 errors, 0 skipped, 0 expected failures, 0 unexpected successes)

--
```

I am building current ToT without this commit to verify if this failure goes away.

ZequanWu commented:
From logs, looks like setting breakpoint with size being 1 byte failed on arm, and works with size being 4 bytes.
